### PR TITLE
ceph: 16.2.10 -> 17.2.5

### PR DESCRIPTION
**Pull request by covert8**
###### Description of changes
Reopening of https://github.com/NixOS/nixpkgs/pull/206396
https://docs.ceph.com/en/latest/releases/quincy/#v17-2-5-quincy

I removed spdk patch because spdk is not used by default. https://github.com/ceph/ceph/blob/v17.2.5/CMakeLists.txt#L266

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/filesystems/ceph/default.nix [hunk 1]
```diff
@@ -140,18 +142,21 @@ let
   ]);
   sitePackages = ceph-python-env.python.sitePackages;
 
-  version = "16.2.10";
+  version = "17.2.5";
   src = fetchurl {
     url = "http://download.ceph.com/tarballs/ceph-${version}.tar.gz";
-    sha256 = "sha256-342+nUV3mCX7QJfZSnKEfnQFCJwJmVQeYnefJwW/AtU=";
+    sha256 = "sha256-NiJpwUeROvh0siSaRoRrDm+C0s61CvRiIrbd7JmRspo=";
```
Comment (mweinelt) [inline pkgs/tools/filesystems/ceph/default.nix:148]:
```suggestion
    hash = "sha256-NiJpwUeROvh0siSaRoRrDm+C0s61CvRiIrbd7JmRspo=";
```

Hash is preferred, as the algorithm is already encoded in the SRI hash.

File: pkgs/tools/filesystems/ceph/default.nix [hunk 2]
```diff
@@ -22,9 +22,12 @@
 , graphviz
 , fmt
 , python39
```
Comment (mweinelt) [inline pkgs/tools/filesystems/ceph/default.nix:24]: Should use python310 ideally, which is the default python version for NixOS 23.05.

**Comment by panicgh** [comment]
Depending on the ceph version, you might try to [unpin](https://github.com/NixOS/nixpkgs/blob/1b90d5e1f9dd0fa39427ad905bb6c351479f8418/pkgs/top-level/all-packages.nix#L6040) it from fmt_8. It may already work with the latest fmt (=fmt_9).

**Comment by mweinelt** [comment]
> Depending on the ceph version, you might try to [unpin](https://github.com/NixOS/nixpkgs/blob/1b90d5e1f9dd0fa39427ad905bb6c351479f8418/pkgs/top-level/all-packages.nix#L6040) it from fmt_8. It may already work with the latest fmt (=fmt_9).

Sadly not with ceph 17.

**Comment by mweinelt** [comment]
@ofborg test ceph-single-node-bluestore
@ofborg test ceph-single-node
@ofborg test ceph-multi-node

File: pkgs/tools/filesystems/ceph/default.nix [hunk 3]
```diff
@@ -125,91 +171,137 @@ let
...
+      # Downgrade rocksdb, 7.10 breaks ceph
+      (rocksdb.overrideAttrs (oldAttrs: {
+        version = "7.9.2";
+        src = fetchFromGitHub {
+          owner = "facebook";
+          repo = "rocksdb";
+          rev = "refs/tags/v7.9.2";
+          hash = "sha256-5P7IqJ14EZzDkbjaBvbix04ceGGdlWBuVFH/5dpD5VM=";
+        };
+      }))
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/ceph/default.nix:282]: Please move that into a let in

File: pkgs/tools/filesystems/ceph/default.nix [hunk 4]
```diff
@@ -1,44 +1,82 @@
...
+, util-linux
+, libuuid
+, udev
+, keyutils
+, rdma-core
+, rabbitmq-c
+, libaio ? null
+, libxfs ? null
+, liburing ? null
+, zfs ? null
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/ceph/default.nix:79]: Disabling dependecies by overwriting them with null is an anti pattern

**Review by SuperSandro2000 — COMMENTED** [review]
We should really do some trick with state version to prevent the automated major update breaking the ceph cluster.

File: pkgs/tools/filesystems/ceph/default.nix (hunk 4 above)
Comment (mweinelt) [reply pkgs/tools/filesystems/ceph/default.nix:79]: Out of scope for this upgrade. I'm not the maintainer, I'm merely making sure we have a supported ceph version in the upcoming release.

File: pkgs/tools/filesystems/ceph/default.nix [hunk 5]
```diff
@@ -226,17 +318,27 @@ in rec {
...
+
+      # Use our own libraries, where possible
+      "-DWITH_SYSTEM_ARROW:BOOL=ON"
+      "-DWITH_SYSTEM_BOOST:BOOL=ON"
+      "-DWITH_SYSTEM_CIMG:BOOL=ON"
+      "-DWITH_SYSTEM_JSONCPP:BOOL=ON"
+      "-DWITH_SYSTEM_GTEST:BOOL=ON"
+      "-DWITH_SYSTEM_ROCKSDB:BOOL=ON"
+      "-DWITH_SYSTEM_UTF8PROC:BOOL=ON"
+      "-DWITH_SYSTEM_ZSTD:BOOL=ON"
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/tools/filesystems/ceph/default.nix:338]: If it is all the same, I'd like to leave it as is.

**Comment by mweinelt** [comment]
> We should really do some trick with state version to prevent the automated major update breaking the ceph cluster.

That would mean maintaining multiple ceph builds, which is expensive, especially when no one maintains them.

**Comment by mweinelt** [comment]
Succeeds all tests on both aarch64-linux and x86_64-linux and should be good to go in.

for x86_64-linux:

```
/nix/store/481g0na8bka1apr9n5bld1ymjyfysi0r-vm-test-run-basic-multi-node-ceph-cluster
/nix/store/33y9ca4ck0w4biqsjzndj1f6mnqq3l00-vm-test-run-basic-single-node-ceph-cluster
/nix/store/inppp8zcm5ca4k4jhyh284z18a5hfkjz-vm-test-run-basic-single-node-ceph-cluster-bluestore
```

**Comment by mweinelt** [comment]
I don't consider this a blocker when it has always been like this. This is an improvement you can certainly invest time in, if you care enough.

I run two ceph clusters, and upgrades from major to the next major version were never a problem for me.

**Comment by mweinelt** [comment]
@ofborg test ceph

**Pull request merged**